components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt

```diff
@@ -289,77 +333,38 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
     /**
      * Check current state then update view state to match.
      */
-    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    fun update() = view?.toScope()?.launch(IO) {
-        val login = Login(
-            guid = guid,
-            origin = origin,
-            formActionOrigin = formActionOrigin,
-            httpRealm = httpRealm,
-            username = username,
-            password = password
-        )
-
-        try {
-            validateStateUpdate?.cancelAndJoin()
-        } catch (cancellationException: CancellationException) {
-            Logger.error("Failed to cancel job", cancellationException)
-        }
-
-        var validateDeferred: Deferred<Result>?
-        validateStateUpdate = launch validate@{
-            val validationDelegate =
-                feature?.loginValidationDelegate ?: return@validate
-            // We only want to fetch this list once. Since we are ignoring username the results will
-            // not change when user changes username or password fields
-            if (potentialDupesList == null) {
-                this@SaveLoginDialogFragment.potentialDupesList = CopyOnWriteArrayList(
-                    validationDelegate.getPotentialDupesIgnoringUsernameAsync(
-                        login
-                    ).await()
-                )
-            }
-            // Passing a copy of this potential dupes list, not a reference, so we know for certain
-            // that what shouldUpdateOrCreateAsync is using can't change.
-            validateDeferred = validationDelegate.shouldUpdateOrCreateAsync(
-                login,
-                potentialDupesList?.toList()
+    private fun update() {
+        // Figure out if we're going to be adding a login or updating an existing one
+        val loginToUpdate = findLoginToUpdate()
+
+        if (loginToUpdate == null) {
+            // Adding a login (or we don't know what's happening because we're
+            // waiting for `loginsStorage.getByBaseDomain()`)
+            setViewState(
+                headline = context?.getString(R.string.mozac_feature_prompt_login_save_headline),
+                negativeText = context?.getString(R.string.mozac_feature_prompt_never_save),
+                confirmText = context?.getString(R.string.mozac_feature_prompt_save_confirmation)
+            )
+        } else {
+            // Updating an existing login
+            setViewState(
+                headline = if (loginToUpdate.username == username) {
+                    // Normal case
+                    context?.getString(R.string.mozac_feature_prompt_login_update_headline)
+                } else {
```

Comment:
This is a different check from before (we were doing isEmpty, not comparing usernames). ISTM that you're assuming that the only way for `loginToUpdate.username != username` is to have `loginToUpdate.username.isEmpty`. What guarantees this assumption? Is it the behaviour of `findLoginToUpdate`? If so, can you document that in a comment? Or simply switch this if statement to what it was before.

E.g. if this assumption doesn't hold you'll be displaying `add_username_headline` when we're actually changing a username on the record. It _is_ possible to change the username for an existing record, right?

Reply:
I believe the only way that this code will change a username is when it fills it in for a login with a blank username (both currently and after the suggested changes).  `findLoginToUpdate` first tries to find a login with a matching username to update.  If it can't find that it tries to find a login with an empty username to fill in.  If neither of those are true, then it returns None and we should add a new login.  I'll add some documentation around this.

Couple of notes here:
  - In the case where the user is saving an empty username and there's an existing login with an empty username, the behavior has changed.  Before it would say use the add username headline, now it uses the update login headline.  I'm pretty sure that's just a bug fix and I checked it with Elise, but tell me if that's not correct.
  - Before the GUID sometimes modified this behavior now it doesn't.